addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/nhc1/NikoHomeControlCommunication1.java

```diff
@@ -0,0 +1,542 @@
+            int port = handler.getPort();
+
+            this.nhcSocket = new Socket(addr, port);
+            Socket socket = this.nhcSocket;
+            this.nhcOut = new PrintWriter(socket.getOutputStream(), true);
+            this.nhcIn = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            logger.debug("Niko Home Control: connected via local port {}", socket.getLocalPort());
+
+            // initialize all info in local fields
+            initialize();
+
+            // Start Niko Home Control event listener. This listener will act on all messages coming from
+            // IP-interface.
+            (new Thread(nhcEvents)).start();
+
+        } catch (IOException | InterruptedException e) {
+            logger.warn("Niko Home Control: error initializing communication");
+            stopCommunication();
+            this.handler.controllerOffline();
+        }
+    }
+
+    /**
+     * Cleanup socket when the communication with Niko Home Control IP-interface is closed.
+     *
+     */
+    @Override
+    public synchronized void stopCommunication() {
+        this.listenerStopped = true;
+
+        if (this.nhcSocket != null) {
+            try {
+                this.nhcSocket.close();
+            } catch (IOException ignore) {
+                // ignore IO Error when trying to close the socket if the intention is to close it anyway
+            }
+        }
+        this.nhcSocket = null;
+
+        logger.debug("Niko Home Control: communication stopped");
+    }
+
+    @Override
+    public boolean communicationActive() {
+        return (this.nhcSocket != null);
+    }
+
+    /**
+     * Runnable that handles inbound communication from Niko Home Control.
+     * <p>
+     * The thread listens to the TCP socket opened at instantiation of the {@link NikoHomeControlCommunication} class
+     * and interprets all inbound json messages. It triggers state updates for active channels linked to the Niko Home
+     * Control actions. It is started after initialization of the communication.
+     *
+     */
+    @SuppressWarnings("null")
+    private Runnable nhcEvents = () -> {
+        @Nullable
+        String nhcMessage;
+
+        logger.debug("Niko Home Control: listening for events");
+        listenerStopped = false;
+        nhcEventsRunning = true;
+
+        try {
+            while (!listenerStopped & ((nhcMessage = this.nhcIn.readLine()) != null)) {
+                readMessage(nhcMessage);
+            }
+        } catch (IOException e) {
+            if (!listenerStopped) {
+                nhcEventsRunning = false;
+                // this is a socket error, not a communication stop triggered from outside this runnable
+                logger.warn("Niko Home Control: IO error in listener");
+                // the IO has stopped working, so we need to close cleanly and try to restart
+                restartCommunication();
+                return;
+            }
+        }
```

Comment:
Maybe put `nhcEventsRunning = false;` in a `finally` block here. To make sure it's always set to false, even on a `RuntimeException`.

Reply:
Done